### DISCUSSION TITLE
A simple start on program repair

### DISCUSSION THREAD

**Discussion by dolfein** [hedyorg/hedy#1225] [Implementation discussions]
The Hedy error messages already provide the users with useful suggestions and hints. As a start on program repair I will focus on the following error messages:
1. Invalid: "{invalid_command} is not a Hedy level {level} command. Did you mean {guessed_command}?"
2. Invalid Space: "Oops! You started a line with a space on line {line_number}. Spaces confuse computers, can you remove it?"
3. Parse: "The code you entered is not valid Hedy code. There is a mistake on line {location[0]}, at position {location[1]}. You typed {character_found}, but that is not allowed."
4. Unquoted Text: "Be careful. If you ask or print something the text should start and finish with a quotation mark. You forgot one somewhere."
5. Var Undefined: "You tried to print the variable {name}, but you did not set it. It is also possible that you were trying to print the word {name} but forgot quotation marks."
6. Unexpected Indentation: "You used too many spaces in line {line_number}. You uses {leading_spaces} spaces, which is too much. Start every new block with {indent_size} spaces."

With number 4 we have to do an (intelligent) guess on where the missing quotation mark should be, this is a little more complicated than just a simple insertion/deletion (especially if the closing one is missing), so I might save this problem for later.
But with the other error messages, the key information is already provided, so it should be easy to repair the program with just simple deletions and/or insertions.

I was thinking of making 2 general functions (in app.js?) to manipulate the editor that can also be used for other stuff later on, something like:
- `insert(row, column, string)` &rightarrow; inserts `string` at position (`row`, `column`)
- `delete(row, column, length)` &rightarrow; deletes `length` characters at position (`row`, `column`)

Each of the above error messages will have a corresponding repair with insertions and/or deletions.

The focus will be mostly on the functionality of those repairs and not so much on the front-end, since there is already a group working on that.

**Comment by Felienne** [comment]
> I was thinking of making 2 general functions (in app.js?) to manipulate the editor that can also be used for other stuff later on, something like:
> 
> * `insert(row, column, string)` → inserts `string` at position (`row`, `column`)
> * `delete(row, column, length)` → deletes `length` characters at position (`row`, `column`)
> 
> Each of the above error messages will have a corresponding repair with insertions and/or deletions.

Great idea, but logic about the programs must live on the Python side of things. I think it would be best to create a new module ProgramRepair in a new file and place this logic there.

A very simple start could be to refactor first, by placing the very limited repair that we already have (where we remove leading spaces) and put that there to se if it all keeps working? That is easier than adding new logic straight away,
